### PR TITLE
[PyTorch] Adjusted the logic of MHA and DPA to enable speculative decoding

### PR DESCRIPTION
**Pull request by Oleg-Goncharov**
This PR modifies the logic of MHA and DPA by using InferenceParams (KV-cache), and extends the supported shapes of matrices in unfused softmax to rectangular to enable speculative decoding.

- Added test cases into `test_numerics.py` to check if the outputs of TransformerLayer or MultiheadAttention generated in a single full run match with those generated incrementally using KV-cache
- The longer the input sequence, the larger the divergence of the output due to accumulation of numerical errors in a forward pass

### CODE REVIEW

File: transformer_engine/pytorch/attention.py [hunk 1]
```diff
@@ -3596,26 +3628,17 @@ def forward(
             batch_start = inference_params.batch_size_offset
             batch_end = batch_start + key_layer.size(1)
             assert batch_end <= inference_key_memory.size(1)
+
             sequence_start = inference_params.sequence_len_offset
             sequence_end = sequence_start + key_layer.size(0)
             assert sequence_end <= inference_key_memory.size(0)
```
Comment (timmoon10) [inline transformer_engine/pytorch/attention.py:3634]: We don't need all this logic here since KV cache logic is handled inside DPA:
```suggestion
            sequence_start = inference_params.sequence_len_offset
            sequence_end = sequence_start + key_layer.size(0)
```

File: transformer_engine/pytorch/attention.py [hunk 2]
```diff
@@ -3596,26 +3628,17 @@ def forward(
...
 
             # adjust the key rotary positional embedding
             if rotary_pos_emb is not None:
+                delta = sequence_end - sequence_start
                 q_pos_emb, k_pos_emb = rotary_pos_emb
-                q_pos_emb = q_pos_emb[sequence_start:sequence_end, :, :, :]
-                k_pos_emb = k_pos_emb[:sequence_end, :, :, :]
+                q_pos_emb = q_pos_emb[:delta, :, :, :]
+                k_pos_emb = k_pos_emb[:delta, :, :, :]
                 rotary_pos_emb = (q_pos_emb, k_pos_emb)
```
[the remark this answers is not in the document]
Comment (ptrendx) [reply transformer_engine/pytorch/attention.py:3637]: I agree with Tim's suggestion - you need to take the embeddings in the right positions for q and k.

For the change in the execution order between RoPE and KV-cache retrieval -> we were discussing it with @ksivaman and @Oleg-Goncharov and we believe this is fine change to make, since in the end KV-cache is just a cache of the values being passed to attention in order to avoid doing the full sequence length every time. It is an optimization of DPA itself about its inputs so it should not matter what those inputs were before (whether they were actually K and V or something completely different). This also makes RoPE logic faster during inference since we are no longer applying it to the fdull sequence every time.

File: transformer_engine/pytorch/attention.py [hunk 3]
```diff
@@ -2643,6 +2646,28 @@ def forward(
         if qkv_format is None:
             qkv_format = self.qkv_format
 
+        if inference_params is not None:
+            assert qkv_format == "sbhd"
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:2650]: Why do we need to assume that? Shouldn't we actually support all cases here (maybe apart from thd, since that would be much harder so we should not tackle it in this PR)? We could inform users that `sbhd` would give the best performance, but `bshd` should also work.

File: transformer_engine/pytorch/attention.py [hunk 4]
```diff
@@ -2643,6 +2646,28 @@ def forward(
...
+            assert self.layer_number is not None
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:2651]: We should have a proper error message here telling the user to set the layer number.

File: transformer_engine/pytorch/attention.py [hunk 5]
```diff
@@ -2643,6 +2646,28 @@ def forward(
...
+            sequence_end = sequence_start + key_layer.size(0)
+            assert sequence_end <= inference_key_memory.size(0)
+
+            # Copy keys and values into KV-cache
+            inference_key_memory[
+                sequence_start:sequence_end, batch_start:batch_end, ...] = key_layer
+            inference_value_memory[
+                sequence_start:sequence_end, batch_start:batch_end, ...] = value_layer
+            key_layer = inference_key_memory[:sequence_end, batch_start:batch_end, ...]
+            value_layer = inference_value_memory[:sequence_end, batch_start:batch_end, ...]
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:2669]: Shouldn't we ensure that K and V are contiguous here? Maybe we make the contiguous later, not sure, but somewhere we need to make sure they are contiguous.

Comment (ptrendx) [reply transformer_engine/pytorch/attention.py:2669]: Since they are not going to be in any other supported format.

File: transformer_engine/pytorch/attention.py [hunk 6]
```diff
@@ -2753,7 +2781,11 @@ def forward(
         if attn_mask_type == "arbitrary":
             use_flash_attention = False
             use_fused_attention = False
-        if "causal" in attn_mask_type and max_seqlen_q != max_seqlen_kv:
+
+        if (inference_params is None
```
Comment (ptrendx) [inline transformer_engine/pytorch/attention.py:2785]: Shouldn't your softmax kernel be able to handle it without inference mode as well?

File: transformer_engine/pytorch/attention.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ptrendx) [reply transformer_engine/pytorch/attention.py:3637]: It would be good to understand why your test worked with the current code since that delta change should change the values.

**Review by timmoon10 — APPROVED** [review]
LGTM, pending CI.
